### PR TITLE
Fix crash when handling event containing NaN or INF values for android

### PR DESCRIPTION
**Pull request by michaelknoch**
## Description
Applies the iOS fix in https://github.com/software-mansion/react-native-reanimated/pull/2896 for android.

Fixes:
- #2662
- #2814
- #2914

## Changes
Fix crash when handling event containing NaN or INF values

## Test code and steps to reproduce
tested by manually inserting NaN and INF values in receiveEvent.java

## Checklist

- [ ] Included code example that can be used to test this change
- [ ] Updated TS types
- [ ] Added TS types tests
- [ ] Added unit / integration tests
- [ ] Updated documentation
- [ ] Ensured that CI passes

### CODE REVIEW

**Comment by michaelknoch** [comment]
cc @tomekzaw @piaskowyk

File: android/src/main/cpp/headers/NativeProxy.h [hunk 1]
```diff
@@ -59,8 +59,16 @@ class EventHandler : public HybridClass<EventHandler> {
       jni::alias_ref<JString> eventKey,
       jni::alias_ref<react::WritableMap> event) {
     std::string eventAsString = "{NativeMap:null}";
-    if (event != nullptr) {
-      eventAsString = event->toString();
+    try {
+      if (event != nullptr) {
```
Comment (tomekzaw) [inline android/src/main/cpp/headers/NativeProxy.h:63]: I'd suggest moving `try/catch` block into `if`.

File: android/src/main/cpp/headers/NativeProxy.h [hunk 2]
```diff
@@ -59,8 +59,16 @@ class EventHandler : public HybridClass<EventHandler> {
...
+        eventAsString = event->toString();
+      }
+    } catch (const std::exception &) {
```
Comment (tomekzaw) [inline android/src/main/cpp/headers/NativeProxy.h:66]:
```suggestion
    } catch (folly::json::parse_error &) {
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]
Thanks for the PR! Indeed, we should handle this on Android analogously as in #2896. BTW Here's the proof that `WritableMap::toString` also uses `folly::toJson` under the hood: https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactAndroid/src/main/jni/react/jni/NativeMap.cpp#L19

I left two suggestions in the comments. Once they are resolved, we are ready to merge I think 😄

File: android/src/main/cpp/headers/NativeProxy.h [hunk 3]
```diff
@@ -60,7 +60,15 @@ class EventHandler : public HybridClass<EventHandler> {
       jni::alias_ref<react::WritableMap> event) {
     std::string eventAsString = "{NativeMap:null}";
     if (event != nullptr) {
-      eventAsString = event->toString();
+      try {
+        eventAsString = event->toString();
+      } catch (std::exception) {
```
Comment (michaelknoch) [inline android/src/main/cpp/headers/NativeProxy.h:65]: apparently `catch(folly::json::parse_error)` doesn't seem to catch the error @tomekzaw

**Comment by michaelknoch** [comment]
thx @tomekzaw , I addressed your feedback

**Review by tomekzaw — APPROVED** [review]
Looks good to me!

**Comment by michaelknoch** [comment]
you think we can get this (and the ios fix) into 2.4.0 release @tomekzaw

**Comment by tomekzaw** [comment]
> you think we can get this (and the ios fix) into 2.4.0 release @tomekzaw

Yes, definitely. These two PRs resolve one of the most mysterious bugs hiding over there. Just waiting to get it double-reviewed (probably tomorrow).

**Review by kmagiera — APPROVED** [review]

**Review by piaskowyk — APPROVED** [review]

**Pull request merged**